client/scala/src/main/scala/ai/verta/blobs/dataset/S3.scala

```diff
@@ -0,0 +1,173 @@
+   */
+  def toVersioningBlob(blob: S3) = {
+    val s3Components = blob.components.map(comp => VersioningS3DatasetComponentBlob(
+      Some(comp),
+      blob.versionMap.get(comp.path.get)
+    ))
+
+    VersioningBlob(
+      dataset = Some(VersioningDatasetBlob(
+        s3 = Some(VersioningS3DatasetBlob(Some(s3Components)))
+      ))
+    )
+  }
+
+  /** Helper function to query metadata of S3 Location
+   */
+  private def getS3LocMetadata(loc: S3Location) = {
+    if (loc.key.isEmpty || loc.key.get.endsWith("/")) {
+      // no key (bucket), or is a folder
+      val versionListing =
+        if (loc.key.isEmpty) s3.listVersions(loc.bucketName, null)
+        else s3.listVersions(loc.bucketName, loc.key.get)
+
+      handleVersionListing(versionListing, List())
+    }
+    else {
+      val request =
+        if (loc.versionID.isDefined) new GetObjectMetadataRequest(loc.bucketName, loc.key.get, loc.versionID.get)
+        else new GetObjectMetadataRequest(loc.bucketName, loc.key.get)
+
+      getObjectMetadata(s3.getObjectMetadata(request), loc.bucketName, loc.key.get).map(List(_))
+    }
+  }
+
+  /** Helper function to deal with VersionListing
+   */
+  @tailrec private def handleVersionListing(
+    versionListing: VersionListing,
+    acc: List[Tuple2[FileMetadata, Option[String]]]
+  ): Try[List[Tuple2[FileMetadata, Option[String]]]] = {
+    val batchAttempt = Try(
+      versionListing.getVersionSummaries().asScala.toList
+                    .filter((version: S3VersionSummary) => !version.getKey().endsWith("/")) // not a folder
+                    .filter(_.isLatest())
+                    .map(getVersionMetadata) // List[Try]
+                    .map(_.get)
+      )
+
+    batchAttempt match {
+      case Failure(e) => Failure(e)
+      case Success(batch) =>
+        if (versionListing.isTruncated())
+          handleVersionListing(s3.listNextBatchOfVersions(versionListing), batch ::: acc)
+        else Success(batch ::: acc)
+    }
+  }
+
+  /** Helper function to extract metadata from the return object
+   */
+  private def getObjectMetadata(obj: ObjectMetadata, bucketName: String, key: String) = Try {
+    val metadata = new FileMetadata(
+      BigInt(obj.getLastModified().getTime()), // convert time to UNIX timestamp (ms)
+      obj.getETag(),
+      getPath(bucketName, key),
+      BigInt(obj.getContentLength())
+    )
+
+    if (obj.getVersionId() != null)
+      new Tuple2(metadata, Some(obj.getVersionId()))
+    else
+      new Tuple2(metadata, None)
+  }
+
+  /** Helper function to extract metadata from a version summary
+   */
+  private def getVersionMetadata(version: S3VersionSummary) = Try {
+    val metadata = new FileMetadata(
+      BigInt(version.getLastModified().getTime()),
+      version.getETag(),
+      getPath(version.getBucketName(), version.getKey()),
+      BigInt(version.getSize())
+    )
+
+    if (version.getVersionId() != null)
+      new Tuple2(metadata, Some(version.getVersionId()))
```

Comment:
I recommend using a quick `case class` instead of tuples, since Tuples are usually associated with key-pair in maps. It makes reading the code much easier, especially when something is a few levels deeper.

Reply:
Will do.

Reply:
Fixed.

Reply:
Updated: Make versionId an optional field of FileMetadata instead. This makes more semantic sense than a separate class + is cleaner to work with.